Mathlib/Tactic/Positivity/Basic.lean

```diff
@@ -38,104 +38,74 @@ end LinearOrder
-    have pa' : Q(by clear! «$pα»; exact 0 < $a) := pa
-    have pb' : Q(by clear! «$pα»; exact $b ≠ 0) := pb
-    pure (.nonzero (q(min_ne_of_lt_of_ne $pa' $pb') : Expr))
-  | .nonzero pa, .positive pb =>
-    have pa' : Q(by clear! «$pα»; exact $a ≠ 0) := pa
-    have pb' : Q(by clear! «$pα»; exact 0 < $b) := pb
-    pure (.nonzero (q(min_ne_of_ne_of_lt $pa' $pb') : Expr))
-  | .nonzero pa, .nonzero pb =>
-    have pa' : Q(by clear! «$pα»; exact $a ≠ 0) := pa
-    have pb' : Q(by clear! «$pα»; exact $b ≠ 0) := pb
-    pure (.nonzero (q(min_ne $pa' $pb') : Expr))
-  | _, _ => pure .none
+  assertInstancesCommute
+  let ⟨_f_eq⟩ ← withDefault do withNewMCtxDepth do assertDefEqQ (u := u.succ) f q(min)
+  match ← core zα pα a, ← core zα pα b with
+  | .positive pa, .positive pb => return .positive q(lt_min $pa $pb)
+  | .positive pa, .nonnegative pb => return .nonnegative q(le_min_of_lt_of_le $pa $pb)
+  | .nonnegative pa, .positive pb => return .nonnegative q(le_min_of_le_of_lt $pa $pb)
+  | .nonnegative pa, .nonnegative pb => return .nonnegative q(le_min $pa $pb)
+  | .positive pa, .nonzero pb => return .nonzero q(min_ne_of_lt_of_ne $pa $pb)
+  | .nonzero pa, .positive pb => return .nonzero q(min_ne_of_ne_of_lt $pa $pb)
+  | .nonzero pa, .nonzero pb => return .nonzero q(min_ne $pa $pb)
+  | _, _ => return .none
 
 /-- Extension for the `max` operator. The `max` of two numbers is nonnegative if at least one
 is nonnegative, strictly positive if at least one is positive, and nonzero if both are nonzero. -/
 @[positivity max _ _] def evalMax : PositivityExt where eval {u α} zα pα e := do
-  let .app (.app f (a : Q($α))) (b : Q($α)) ← withReducible (whnf e) | throwError "not max"
-  let result ← catchNone do
-    let _a ← synthInstanceQ (q(LinearOrder $α) : Q(Type u))
-    guard <|← withDefault <| withNewMCtxDepth <| isDefEq f q(max (α := $α))
+  let .app (.app (f : Q($α → $α → $α)) (a : Q($α))) (b : Q($α)) ← withReducible (whnf e)
+    | throwError "not max"
+  let _e_eq : $e =Q $f $a $b := ⟨⟩
+  let _a ← synthInstanceQ (q(LinearOrder $α) : Q(Type u))
+  assertInstancesCommute
+  let ⟨_f_eq⟩ ← withDefault do withNewMCtxDepth do assertDefEqQ (u := u.succ) f q(max)
+  let result : Strictness zα pα e ← catchNone do
     let ra ← core zα pα a
     match ra with
-    | .positive pa =>
-      have pa' : Q(by clear! «$pα»; exact 0 < $a) := pa
-      pure (.positive (q(@lt_max_of_lt_left $α _ _ _ $b $pa') : Expr))
-    | .nonnegative pa =>
-      have pa' : Q(by clear! «$pα»; exact 0 ≤ $a) := pa
-      pure (.nonnegative (q(@le_max_of_le_left $α _ _ _ $b $pa') : Expr))
+    | .positive pa => return .positive q(lt_max_of_lt_left $pa)
+    | .nonnegative pa => return .nonnegative q(le_max_of_le_left $pa)
     -- If `a ≠ 0`, we might prove `max a b ≠ 0` if `b ≠ 0` but we don't want to evaluate
     -- `b` before having ruled out `0 < a`, for performance. So we do that in the second branch
     -- of the `orElse'`.
-    | _ => pure .none
+    | _ => return .none
```

Comment:
Similarly I actively avoid using `return` when it's not an early return, especially in a case like this with a nested `do` block.